<a href="https://colab.research.google.com/github/ksg7758/receipt-scanner/blob/master/Katherine_Graci_A1_Python_and_Pandas_AI.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Assignment 1 — Python, pandas, and Data That Fights Back


Graded on completion. You may work with others; each of you submits your own copy.

---

**Your name / NYU email:** Katherine Graci, ksg7758@stern.nyu.edu

**Worked with:**

---

This is the warm-up. It exists because the rest of the term assumes you can get
data into Python and trust what comes out.

**Use AI for all of it.** Every line of code here is something Gemini or ChatGPT
will write for you in seconds, and you should let it. What the AI will *not* do
is notice when the answer is wrong — and this assignment is mostly a tour of
ways a notebook returns a confident, plausible, incorrect number.

Where a question asks you to **predict before running**, do that honestly. The
prediction is the point; getting it wrong costs you nothing.

---

# Part 1 — Two ways a notebook will lie to you

### Q1 — Cells run in the order *you* run them, not the order they appear

Run the next two cells in order. You should get 2,680.19.

In [5]:
deposit = 2000
rate    = 0.05
years   = 6

In [6]:
deposit * (1 + rate) ** years

2680.191281250001

Now, in the cell below, change the rate to **10%** — and do *not* re-run
anything above.

In [3]:
rate = 0.10

Go back and re-run **only** the `deposit * (1 + rate) ** years` cell.

> **(a)** What number do you get now, and why is it not the 5% answer?
>
> **(b)** Now scroll up and re-run the *first* cell, then the calculation cell
> again. What happens, and why?
>
> **(c)** A notebook has an execution *order* and a visual *order*, and they are
> not the same thing. Describe, in one or two sentences, how this could produce a
> number in your final report that you cannot reproduce the next morning.

This is the single most common way a Jupyter result turns out to be wrong, and
AI makes it worse rather than better — pasting a fix halfway up the notebook and
re-running one cell is exactly the move that causes it.

**(a)** I get 3,543.12, and it is not the 5% answer because changing the rate makes the answer larger. In that cell, we are calculating future rate, which increases as rate increases, so that is why the 5% and 10% answers are different with the 10% one being larger.

**(b)** It goes back to the original 2,680.19 answer because the rate was set to 5% again.

**(c)** Execution order is the sequence in which we run the code, which may not line up with visual order which is the physical arrangement of cells from top to bottom on the page. This could produce a number that is incorrect if I accidentally run a cell that affects variables' values from the end of the page before running the entire document.

### Q2 — Code that runs, returns a plausible number, and is wrong

You put \$100 into a stock. It returns **+15%** in year one and **−14%** in
year two. The cell below claims to compute what you end up with.

**Do not fix it yet.** First: look at it, and write down what you think it will
print.

In [7]:
r1 = 0.15
r2 = -0.14

100 * 1 + r1 * 1 + r2      # <- what will this print?

100.01

> **(a)** Your prediction, before running:
>
> **(b)** Run it. What did it actually print? Is that number plausible as
> "\$100 after two years"? Would you have caught it if it had appeared in the
> middle of a table?
>
> **(c)** Fix it in the cell below, and state the correct final value.

In [8]:
# Your corrected calculation
100 * (1 + r1) * (1 + r2)

98.89999999999999

**(a)** prediction: My prediction is that it will print a number slightly lower than 100.  

**(b)** It actually printed a number slightly above 100. It is not plausible because after growing 15%, 14% the following year would be a larger dollar amount and the stock's value should be below 100. I probably wouldn't have caught it in the middle of a table.

**(c)** correct value: 98.90

---

# Part 2 — Python worth knowing when the AI writes the code

### Q3 — Predict, then run

Below are six statements. `x = 2`, `y = 2`, `z = 4`.

```python
x > z                              # 1
x == y                             # 2
(x < y) and (x > y)                # 3
(x < y) or  (x > y)                # 4
(x <= y) and (x >= y)              # 5
True and ((x < z) or (x < y))      # 6
```

**Write down your six answers first**, as a list like `[True, False, ...]`.
Then run them and compare. Note any you got wrong — `and`/`or` precedence is a
real source of silently wrong filters later in the course.

**My predictions:** `[False,True,False,False,True,True]`

In [13]:
x, y, z = 2, 2, 4
# check your six predictions here
print(f"1: {x > z}")
print(f"2: {x == y}")
print(f"3: {(x < y) and (x > y)}")
print(f"4: {(x < y) or  (x > y)}")
print(f"5: {(x <= y) and (x >= y)}")
print(f"6: {True and ((x < z) or (x < y))}")

1: False
2: True
3: False
4: False
5: True
6: True


### Q4 — Data does not arrive as numbers

You are handed a price as `"$6.50"`. Python sees a string, and `"$6.50" * 2`
gives you `"$6.50$6.50"` rather than 13.

Turn it into the float `6.5`. Then say in one sentence what would happen if a
column of a thousand prices had this problem and you took its mean.

In [16]:
price = "$6.50"
# your code here
price_float = float(price.replace('$', ''))
print(price_float)

6.5


**What happens to the mean of a column like this:**

### Q5 — Why logs show up everywhere in finance

There is a trick worth knowing: for numbers close to 1, the percent change
$(x-y)/y$ is well approximated by the difference in logs, $\log x - \log y$.

Verify it with the numbers below — compute both and compare. Then try it again
with `x = 2.0, y = 1.0` and report how well the approximation holds.

One sentence: when is it safe to use, and when is it not?

In [17]:
import math

x, y = 1.05, 1.02

# Calculate percent change
percent_change_1 = (x - y) / y

# Calculate difference in logs
diff_logs_1 = math.log(x) - math.log(y)

print(f"For x={x}, y={y}:")
print(f"  Percent Change: {percent_change_1:.6f}")
print(f"  Difference in Logs: {diff_logs_1:.6f}")

x_large, y_large = 2.0, 1.0

# Calculate percent change for larger difference
percent_change_2 = (x_large - y_large) / y_large

# Calculate difference in logs for larger difference
diff_logs_2 = math.log(x_large) - math.log(y_large)

print(f"\nFor x={x_large}, y={y_large}:")
print(f"  Percent Change: {percent_change_2:.6f}")
print(f"  Difference in Logs: {diff_logs_2:.6f}")

For x=1.05, y=1.02:
  Percent Change: 0.029412
  Difference in Logs: 0.028988

For x=2.0, y=1.0:
  Percent Change: 1.000000
  Difference in Logs: 0.693147


When `x` and `y` are close to 1 (e.g., 1.05 and 1.02), the percent change and the difference in logs are very similar, indicating a good approximation. However, when `x` and `y` are far apart (e.g., 2.0 and 1.0), the approximation breaks down, and the values are no longer close.

---

# Part 3 — Real data, which does not want to help you

Everything from here uses one Excel file: 49 US industry portfolios from Ken
French, monthly, going back to 1926, plus a sheet with the market return and the
risk-free rate.

```
https://raw.githubusercontent.com/amoreira2/UG54/refs/heads/main/assets/data/Assignment1.xlsx
```

**Download it and open it in Excel before you write any code.** Ten seconds of
looking will save you an hour. This is a habit, not a suggestion.

In [18]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
plt.style.use('seaborn-v0_8-whitegrid')
plt.rcParams['figure.figsize'] = [10, 4]

URL = ("https://raw.githubusercontent.com/amoreira2/UG54/"
       "refs/heads/main/assets/data/Assignment1.xlsx")
pd.ExcelFile(URL).sheet_names

['Market_proxy', '49_Industry_Portfolios']

### Q6 — Load it, and count the columns

Read the `49_Industry_Portfolios` sheet. The real column names are **not** in
the first row — there are several lines of description above them, so you will
need `skiprows`.

Then answer this before going further: **how many columns did you get?** I got 49 columns

> **(a)** How many columns are there, and how many industries were you promised?
>
> **(b)** Print the column names. Some of them end in `.1` and `.2`. What has
> happened, and what would you have computed if you had just averaged everything?
>
> **(c)** Keep only the block you actually want — the value-weighted returns —
> and say how you know it is the right one.

In [21]:
# your code here
import pandas as pd

# URL is already defined in the notebook context
# URL = ("https://raw.githubusercontent.com/amoreira2/UG54/"\
#        "refs/heads/main/assets/data/Assignment1.xlsx")

# Read the '49_Industry_Portfolios' sheet, skipping the first 5 rows
raw = pd.read_excel(URL, sheet_name='49_Industry_Portfolios', skiprows=6)
print(raw.shape)
print(list(raw.columns))
# Deduce how many columns there are
num_columns = raw.shape[1]
print(f"Total number of columns: {num_columns}")

# Print the column names
print("All column names:", list(raw.columns))

# Keep only the first block which is value weighted returns
# Assuming the 'Date' column is the first, followed by 49 industry columns
# So, we select the first 50 columns (index 0 to 49)
ind = raw.iloc[:, :50]

print(f"\nShape of the value-weighted industries DataFrame: {ind.shape}")
print("Value-weighted industry column names:", list(ind.columns))

(1069, 152)
['Unnamed: 0', 'Agric', 'Food', 'Soda', 'Beer', 'Smoke', 'Toys', 'Fun', 'Books', 'Hshld', 'Clths', 'Hlth', 'MedEq', 'Drugs', 'Chems', 'Rubbr', 'Txtls', 'BldMt', 'Cnstr', 'Steel', 'FabPr', 'Mach', 'ElcEq', 'Autos', 'Aero', 'Ships', 'Guns', 'Gold', 'Mines', 'Coal', 'Oil', 'Util', 'Telcm', 'PerSv', 'BusSv', 'Hardw', 'Softw', 'Chips', 'LabEq', 'Paper', 'Boxes', 'Trans', 'Whlsl', 'Rtail', 'Meals', 'Banks', 'Insur', 'RlEst', 'Fin', 'Other', 'Unnamed: 50', 'Unnamed: 51', 'Agric.1', 'Food.1', 'Soda.1', 'Beer.1', 'Smoke.1', 'Toys.1', 'Fun.1', 'Books.1', 'Hshld.1', 'Clths.1', 'Hlth.1', 'MedEq.1', 'Drugs.1', 'Chems.1', 'Rubbr.1', 'Txtls.1', 'BldMt.1', 'Cnstr.1', 'Steel.1', 'FabPr.1', 'Mach.1', 'ElcEq.1', 'Autos.1', 'Aero.1', 'Ships.1', 'Guns.1', 'Gold.1', 'Mines.1', 'Coal.1', 'Oil.1', 'Util.1', 'Telcm.1', 'PerSv.1', 'BusSv.1', 'Hardw.1', 'Softw.1', 'Chips.1', 'LabEq.1', 'Paper.1', 'Boxes.1', 'Trans.1', 'Whlsl.1', 'Rtail.1', 'Meals.1', 'Banks.1', 'Insur.1', 'RlEst.1', 'Fin.1', 'Other.1

**(a)** There are 152 columns, but they only include the 49 industries we were promised. They repeat for different statistics.

**(b)** Some of them end in .1 or .2 because they are repeats of a column that already exists but within a block for another statistic. If we had just averaged everything, we would have been combining different computations together.

**(c)** We know it's the block we want because I manually checked to see in the spreadsheet that the first 50 columns are the value weighted returns.

In [23]:
# Keep only the date column ('Unnamed: 0') and the 49 value-weighted industry columns
# These correspond to the first 50 columns (index 0 to 49)
ind = raw.iloc[:, :50]

print(f"Shape of the value-weighted industries DataFrame: {ind.shape}")
print("Value-weighted industry column names:", list(ind.columns))

Shape of the value-weighted industries DataFrame: (1069, 50)
Value-weighted industry column names: ['Unnamed: 0', 'Agric', 'Food', 'Soda', 'Beer', 'Smoke', 'Toys', 'Fun', 'Books', 'Hshld', 'Clths', 'Hlth', 'MedEq', 'Drugs', 'Chems', 'Rubbr', 'Txtls', 'BldMt', 'Cnstr', 'Steel', 'FabPr', 'Mach', 'ElcEq', 'Autos', 'Aero', 'Ships', 'Guns', 'Gold', 'Mines', 'Coal', 'Oil', 'Util', 'Telcm', 'PerSv', 'BusSv', 'Hardw', 'Softw', 'Chips', 'LabEq', 'Paper', 'Boxes', 'Trans', 'Whlsl', 'Rtail', 'Meals', 'Banks', 'Insur', 'RlEst', 'Fin', 'Other']


In [24]:
print(ind.shape)

(1069, 50)


### Q7 — Missing values that are not missing

The header text says: *"All missing values are indicated by -99.99 or -999."*

Find them and replace them with proper `NaN`. Then report what fraction of the
table was missing.

> **⚠️** If you skip this, `-99.99` is a perfectly valid number and every mean,
> standard deviation and correlation you compute will quietly include it as a
> **−99.99% monthly return**. Nothing will error.

In [25]:
import numpy as np

# Replace -99.99 and -999 with NaN
ind_cleaned = ind.replace([-99.99, -999], np.nan)

# Calculate total number of cells
total_cells = ind_cleaned.size

# Calculate number of missing values
missing_values = ind_cleaned.isna().sum().sum()

# Calculate fraction of missing values
fraction_missing = missing_values / total_cells

print(f"Total cells in the table: {total_cells}")
print(f"Number of missing values: {missing_values}")
print(f"Fraction of the table that was missing: {fraction_missing:.4f}")

Total cells in the table: 53450
Number of missing values: 2877
Fraction of the table that was missing: 0.0538


### Q8 — Dates and units

Two more things stand between you and usable data.

**Dates.** The first column is an integer like `192607`. Convert it to a proper
month-end date and make it the index. Month-*end* matters: these are monthly
returns, and later in the course you will merge them against other series that
are stamped at month end. A one-day mismatch silently drops every row.

**Units.** Print the mean of one industry. Is it a monthly return of 0.9%, or
90%? Convert so that a 1% return is stored as `0.01`.

In [26]:
# Convert 'Unnamed: 0' to datetime and set as index
ind_cleaned['Date'] = pd.to_datetime(ind_cleaned['Unnamed: 0'], format='%Y%m') + pd.offsets.MonthEnd(0)
ind_cleaned = ind_cleaned.set_index('Date')
ind_cleaned = ind_cleaned.drop(columns=['Unnamed: 0'])

# Print the index to verify date conversion
print("First 3 dates in the index:", ind_cleaned.index[:3])

# Print mean of 'Agric' before unit conversion (still in percent)
print(f"Mean monthly return, Agric (in percent): {ind_cleaned['Agric'].mean():.5f}")

# Convert all return columns from percent to decimal (divide by 100)
# Exclude any non-numeric columns if they were to exist, though not the case here
ind_cleaned.iloc[:, :] = ind_cleaned.iloc[:, :] / 100

# Print mean of 'Agric' after unit conversion (in decimal)
print(f"Mean monthly return, Agric (in decimal): {ind_cleaned['Agric'].mean():.5f}")

First 3 dates in the index: DatetimeIndex(['1926-07-31', '1926-08-31', '1926-09-30'], dtype='datetime64[ns]', name='Date', freq=None)
Mean monthly return, Agric (in percent): 0.96393
Mean monthly return, Agric (in decimal): 0.00964


### Q9 — Excess returns

Load the `Market_proxy` sheet the same way — it has `Mkt-RF` and `RF`. Watch the
units here too.

Then build excess returns two ways:

1. For **one** industry, `Agric`, subtract `RF` month by month. Print its mean.
2. For **all 49 at once**, in a single line, producing a DataFrame called `inde`.

> **🤖 Worth asking the AI:** *"subtract a Series from every column of a
> DataFrame, aligning on the index"* — and then check that the row count did not
> change. Getting the axis wrong here produces a table full of NaN, or worse, a
> table that looks fine and is transposed.

In [31]:
# Load the 'Market_proxy' sheet, explicitly setting header to the 4th row (index 3)
market = pd.read_excel(URL, sheet_name='Market_proxy', header=3)

# The current DataFrame 'market' has an unexpected name for the first column (the date column).
# It is named 'The 1-month TBill return is from Ibbotson and Associates Inc.'
# We need to rename this column to 'Date'.
market = market.rename(columns={market.columns[0]: 'Date'})

# Clean market data similar to industries
# Convert the 'Date' column to datetime and set it as the index
market.index = pd.to_datetime(market['Date'], format='%Y%m') + pd.offsets.MonthEnd(0)
market = market.drop(columns=['Date'])

# Replace missing values (if any) and convert to decimal
# The Mkt-RF and RF columns might also contain -99.99 or -999, so apply replacement to all relevant columns
market.iloc[:, :] = market.iloc[:, :].replace([-99.99, -999], np.nan)
market.iloc[:, :] = market.iloc[:, :] / 100


# Ensure both DataFrames have the same date range and columns aligned
# We will align ind_cleaned and market on their common index

# Extract RF (Risk-Free rate)
RF = market['RF']

# 1. For one industry, Agric, subtract RF month by month
ex_ret_agric = ind_cleaned['Agric'] - RF
print(f"Mean excess return for Agric: {ex_ret_agric.mean():.5f}")

# 2. For all 49 at once, in a single line, producing a DataFrame called inde
# We need to ensure that the RF Series is broadcast correctly across the DataFrame.
# Pandas automatically aligns by index, so simple subtraction works.
inde = ind_cleaned.subtract(RF, axis=0)

print(f"\nShape of the excess returns DataFrame (inde): {inde.shape}")
print("First 3 rows of excess returns for a few industries:")
print(inde.head(3))

KeyError: 'Date'

### Q10 — Dropping rows has a price

Not every industry exists in 1926. To get a rectangular table where every
industry covers the same months, drop the incomplete rows.

Report: how many months you had before, how many after, and **what date the
sample now starts**.

> **(a)** How many months did that cost you?
>
> **(b)** You have just thrown away 40 years of data to keep 49 columns. Name
> one thing you might have done instead, and say what it would have cost.

In [ ]:
# your code here


**(a)**

**(b)**

### Q11 — The two objects everything else is built from

Compute and display:

1. `ERe` — the vector of **annualized** mean excess returns, one per industry
2. `CovRe` — the **annualized** covariance matrix of excess returns

> **If you have not met a covariance matrix.** For 49 industries it is a 49×49
> table. The diagonal entry $(i,i)$ is industry $i$'s **variance** — its
> volatility squared. The off-diagonal entry $(i,j)$ is the **covariance**
> between $i$ and $j$: positive when the two tend to move in the same direction,
> negative when one tends to rise as the other falls, near zero when neither
> tells you much about the other.
>
> A **correlation** is that same covariance divided by the two volatilities,
> which rescales it to sit between −1 and +1. Same information, easier to read —
> which is why the question below asks you for correlations rather than raw
> covariances.
>
> `df.cov()` builds the whole table in one call. Nothing in the next few weeks
> needs it. It comes back in Capital Allocation in October, where it is the
> object that decides how much of each thing to hold.

Report the highest and lowest average-return industries. Then print the
correlation between three pairs of your choosing — pick one pair you expect to
move together and one you expect not to, and say whether the data agreed.

Remember: means annualize by ×12, variances by ×12, volatilities by ×√12.

In [ ]:
# your code here


**Pairs I picked, what I expected, what I found:**

### Q12 — Look at it

Pick **two** industries that you think behaved very differently in some
identifiable episode — a crisis, a boom, a technological shift.

1. Plot their monthly excess returns over that period.
2. Plot the **cumulative** return of \$1 invested in each over that period.
3. In three or four sentences, say what happened, using the actual magnitudes
   from your plot. What would \$1 have become in each?

> **📌** Cumulative returns compound: `(1 + r).cumprod()`, not `r.cumsum()`.
> Adding returns is Lecture 1's pitfall 4 and the error grows with horizon.

In [ ]:
# your code here


**What happened:**

---

## 📤 Submission

1. **Run your notebook from a clean start** — Runtime → Restart and run all.
   If it does not survive that, Q1 is still ahead of you.
2. File → Download → `.ipynb`
3. Upload to Brightspace under Assignment 1.

---

## What this was for

Next week you meet a dataset with 1.5 million rows where you cannot open the
file and look. Every trap in Part 3 is one you will hit again there, invisibly:
columns that are not what they claim, sentinel values masquerading as data,
dates that do not line up, and percent where you assumed decimal.

You will also notice that this file is **wide** — one column per industry. The
course panel is **long** — one row per stock-month. Lecture 2 explains why, and
having built something in the wide shape first is the point.